### ISSUE TITLE
Feature request: Python SDK

### ISSUE THREAD

**Issue by whisper-bye** [slatedb/slatedb#243]
As the title.

**Comment by criccomini** [comment]
I believe @gunnarmorling is working on an FFI Java binding.

**Comment by gunnarmorling** [comment]
Yeah, I am doing some explorations right now. The issue I ran in so far is that not all the types are FFI-safe, for instance there are structs in API methods which have a field of type `time::Duration`. For the types of the project itself you can ensure a stable ABI by means of annotating them, but IIUC, that's not possible for externally defined types like `Duration`. So it seems I'd have to create a separate FFI-compatible API which only uses FFI-safe types (e.g. either using `long` instead of `Duration`, or a project-specific `Duration` type, or we'd have to change the current SlateDB APIs to conform to these requirements. Gonna continue with my experiments as I have time, also just learning the basics of Rust along the way :)

**Comment by criccomini** [comment]
I feel like changing SlateDB to make it FFI safe is the right way to go. Is it just duration?

**Comment by gunnarmorling** [comment]
> I feel like changing SlateDB to make it FFI safe is the right way to go.

👍 

> Is it just duration?

Probably more, but I need to actually go through the API and see. So `DB` is the main entry point, right? Is there any others besides it?

**Comment by criccomini** [comment]
Db is the main entry point. Eventually, we want to allow users to start the compactor and GC as separate processes, but that's off in the distance.

**Comment by batmilkyway** [comment]
Most rust -> python bindings use https://www.maturin.rs/ from what I can tell, see https://github.com/delta-io/delta-rs/tree/main/python for example

**Comment by KaguraMilet** [comment]
We might be able to refer to bindings of Opendal.
[Opendal bindings](https://github.com/apache/opendal/tree/main/bindings)

**Comment by dsgibbons** [comment]
I'm happy to take a look at this. I note that there already exists Python bindings for `object_store`: https://github.com/roeap/object-store-python, which should make things easier.

**Comment by dsgibbons** [comment]
@criccomini after spending a bit of time on this, I found that this was more involved than I first expected. I probably don't have the time or PyO3 expertise to do this at the moment. However, I'd encourage anyone who is interested to look into [object-store-rs](https://github.com/developmentseed/object-store-rs) for a good example of using PyO3 to generate bindings for `object_store`, which may prove to be a good starting point.

**Comment by vigneshc** [comment]
@gunnarmorling , any outcome from the exploration that you share on java binding?

**Comment by gunnarmorling** [comment]
Also just logged #368 for the Java bindings specifically.

**Comment by dmvk** [comment]
Hey @vigneshc @gunnarmorling, unfortunately I currently don't have capacity to do contributions in this direction.

For an internal use-case, I have handwritten use-case specific bindings (based on `jni-rs`), that don't and won't cover the whole surface area that is needed for general purpose SDK. It's also in very early stage to share any meaningful insight about the performance.

**Comment by criccomini** [comment]
A basic SDK has been merged in #621. I'm going to close this and open GH issues for specific features that need to be added to the SDK.
